### ISSUE TITLE
CSV-report - Timeline

### ISSUE THREAD

**Issue by M0nk3yOo** [CompliantSec/M365SAT#37]
Dear asterictnl-lvdw,

do you have any expectation when the CSV-report will be included?
I see the hint "in the next release" since a few releases back.

(I don't wanna push you, just for my/our information, because we (ps-tron and me) would really appreciate it :-) )

Best
M0nk3y

**Comment by asterictnl-lvdw** [comment]
Good morning,

Yeah, it has been a bit more work than usual since the structure of translating the object and displaying it correctly in a .csv file did not kinda work properly. Also, there were other features that had more priority over it. But I will promise that I will put it in the next release or in a quick bugfix release, depending on what will be changed this version. 

I might release it to Alpha so you can test it out there to see if it's what you are expecting at the moment.

**Comment by M0nk3yOo** [comment]
Thanks for your fast reply, that would be great! :-)

**Comment by M0nk3yOo** [comment]
May you put the "pass" in the CSV-export as well, please?
(and maybe an information from those, which could not be checked and needs to be done manually?)

**Comment by asterictnl-lvdw** [comment]
Hoi,

Could you give a little bit more context about this? It would be great to determine which information you want in the .CSV to see if this information is important for an audit for example or not.

Thanks in advance.

~Leonardo

**Comment by M0nk3yOo** [comment]
Dear Leonardo,

sorry for my latency, I was to busy...

The idea is to automate the CSV-report as much as possible, based on the CIS-report Excel sheet.
Here an example:

![image](https://github.com/asterictnl-lvdw/M365SAT/assets/32765291/f85b3c5f-cae6-476d-8a7f-3801248a31c7)

In best case it would include the automated and the manual checks as well. And if the check ends in an error or is not possible to check automatically, it should have the corresponding type.


Have a nice day! :)

**Comment by asterictnl-lvdw** [comment]
Hello @M0nk3yOo ,

Thank you for the submission. This would clearly be two enhancements that have some relation to each other.

What I am going to do is the following:
- I am going to automate the scripts where possible that are manually checked on the first hand. Some of the things could be automated via scripts. But I did not find everything yet. I need some more research about that. 
- I am planning once v4.0.0 of the Microsoft 365 benchmark comes out to review it once again and to implement the changes and also to see what manual checks need to be implemented alongside to fully automate things. 
- The same is for the Azure benchmark as I want to wait until that release is done as well. I want to change some things around and implement it back to the CIS v3.1.0, v3.0.0 and v2.0.0 and maybe even the v1.x.x if I have the time

Currently I haven't put all the information in and only the necessary information. The script is there for to allow checking upon CIS checkpoints to determine if settings are compliant or not and not for actual CIS-reporting at this moment. 
- I will look into adding the OK status as well with a score of 0 or OK depending on what I find better looking for the report. This will make the report clearer and users can see what they already have compliant and what they need to fix.
The other thing about the .csv I will push something to Alpha once I am done and I will allow you some testing to see if it's something that is upon your liking as well.

~LvdW

**Comment by M0nk3yOo** [comment]
Hi @asterictnl-lvdw 

many thanks for your detailed explanation! :-)
I really appreciate it.

Maybe it makes sense, even if not all checks can be done automatically, to have them in the report/csv(?)


Best

**Comment by asterictnl-lvdw** [comment]
Can you test the alpha branch for me @M0nk3yOo ? I have submitted the .csv module and I want to check if it has the things that are required and of course if it works.

I will test it myself later as well. But I am busy developing and looking at some things as well.

~Leonardo

**Comment by M0nk3yOo** [comment]
Hmm... I'm having an issue while running the script(s).
<img width="1028" alt="image" src="https://github.com/asterictnl-lvdw/M365SAT/assets/32765291/75f80a53-241b-40a5-841a-f671f4d4ec57">

This is while connecting to Graph (Connect-MgGraph). Do you have a hint for me where the permission change is needed? I remember that there is a registration/permission assignes while running the script the first times, but now I'm having this issue and have no clue (if I'm serious) which app should have the additional permission

**Comment by asterictnl-lvdw** [comment]
Ah, this is related to an implementation in #34 I will remove the Policy.Read.DeviceConfiguration because it seems that it is no such permission there. I will push another alpha release with that fix. 

~LvdW

**Comment by asterictnl-lvdw** [comment]
Update: You can try and pull the latest version of Alpha now. It should have been fixed by now so you will get no error by logging in anymore.

~LvdW

**Comment by M0nk3yOo** [comment]
The issue itself is solved :-)
(I think - I deleted some references and apps from my profile, don't know if it works because of your change, or what I did. But I think it's more your fix :-) )

Topic CSV: The CSV is created succsessfully, but the result is (today) not in that format (and its content), what we 100% think. But I think the baseline ha a very good stand.
Maybe, if I have some time (trying to reach some MS-certificates such as SC-200), will create some additional CIS-checks.

**Comment by asterictnl-lvdw** [comment]
That's good! Could you tell me what you are missing so I can see if I can add it and if it is valuable to add to the CSV report.

Feedback is always welcome!

~LvdW

**Comment by M0nk3yOo** [comment]
Hi, I just want give a "hello" from my side, I'm still here, but still had no time to work on that topic, but I will (but don't know when)!

Best

**Comment by M0nk3yOo** [comment]
Dear @asterictnl-lvdw 

I'm currently checking the code. Even if I'm to stupid to debug the code in VSCode, maybe some useful changes for us:
- Add a new attribute in the inspectors for the particular checks. For example for ID CISMAz5111
`ID			     = "CISMAz5111"`
additional ID-attribute, mapped to the CIS number as like in the "FindingName"
`ID2 = "5.1.1.1"`
This might be easier to lookup the controls in an PDF document (unfortunately I can't poste it here from license reason)

- Add the different level as additional attribute to sort/filter afterwards in CSV file
- List all results, not only the failed ones to the "$findings" array and add an additional parameter "Result" with "failed/pass". This should bring the CSV report to a complete audit result. Checks run into in error, should also be listed and marked as "errored" or sth. like that
- Some "ReturnedValue"s have been reportet as PS-object.

If you give me a hint how I can debug the code, I can try to bring in the changes from my side (and maybe another column sort)


In addition I'll ask my boss if we can give you donation for your work, which save us some time to audit our customers :)

#############

EDIT: Debug is working, I started the wrong file, sorry!

**Comment by asterictnl-lvdw** [comment]
Dear @M0nk3yOo ,

Sorry for the late reaction. 

I will process your feedback in the alpha branch and will give a reply here once you are able to test. 

I'll add the CIS-ID in the CSV report. To achieve this, I have to make changes in the inspector modules.

Don't worry, because I have the controls as well so I can easily look them up and the ID is processed in the finding name as well.

For the Level and license, I have to make additional attributes in the PSObject in each inspector as well. This is also handy for the HTML report so I will process this as well in the alpha branch.

I have been looking at the PASS/FAIL schema on how to implement this as this requires code changes as I can add this either in the validation process and pass this as a parameter into each Build function or I can do this a different way. 

If I am going to implement the PASS/FAIL structure I am going to create a formula to determine the risk-score as the maximum risk score is determined onto the number of inspectors being ran at that point. 

I need more time onto the PASS/FAIL as this depends on multiple factors within the code which requires code change in additional places and not only the inspector modules. I will hopefully bring this feature up to the next major release including with new things. 

Donations are very much appreciated <3

I hope I have informed you in proper manner. :)

~LvdW

**Comment by olivierdumon** [comment]
Hi @asterictnl-lvdw 

Is it possible to have an update concerning this request ?
I am also very interested by the possibility of creating reports in CSV format.

Thank you in advance for your reply.
and congrats for this tool

Good day
Olivier

**Comment by asterictnl-lvdw** [comment]
Hello @olivierdumon ,

Thank you for your reply.

The optimalizations regarding the .CSV-format are planned for the next major release. 

Since with CSV you want the information to be as simplified as possible, I am looking at besides the feedback which was given earlier to look on how to optimally implement it altogether which benefits not only the .CSV file but also the .HTML file, because there are multiple things are gone well together and could be implemented simultaneously. 

I hope to a tiny release at the beginning of Q4 (September) and hopefully somewhere before the end of this year, maybe even sooner the major release including optimalisations in both the CSV and the whole reporting structure. 

I hope this gives you a brief overview on what to expect.

This to let you know that it is on the planning, but it takes more time to process and there are still some important bugs to fix first and to release the v3.1 of the CIS first before doing optimalisations on the whole CSV exporting engine.

~LvdW

**Comment by olivierdumon** [comment]
Hi @asterictnl-lvdw 

Thank you for your feedback

I'll be following your project closely :)
Thank you for your work

**Comment by M0nk3yOo** [comment]
Hi @asterictnl-lvdw,

regarding the donation: is there a way to get in direct contact to you? Maybe via mail?


Thanks, and don't melt away today :-)

**Comment by asterictnl-lvdw** [comment]
Hi @M0nk3yOo ,

On my profile there should be a LinkedIn of mine. You can send me a message there.

~LvdW

**Comment by asterictnl-lvdw** [comment]
Hi everyone,

Just to give a quick update about this. I am working on version 3.0 which will hopefully include the CSV support including the additional feedback above implemented. It takes more time than planned, because there need to be changes into the way of creating Inspectors as well for the .CSV lay-out. 

~LvdW

**Comment by asterictnl-lvdw** [comment]
@M0nk3yOo , @olivierdumon , 

How does this look?

![image](https://github.com/user-attachments/assets/6c141465-9d56-4f7c-a8b6-a85e586a9815)


Is this sufficient, because I need to fix all the scripts. This is important for me to know.

I will reset the titles maybe to something different, depending on what the context would be.

~LvdW

**Comment by M0nk3yOo** [comment]
Hi Leonard,

I think this looks very good. I think "Status" includes the fail/pass result. This should be available in the CSV, yes! :-)
For debugging/evidence purpose it might also be nice, if the returned and expected vales are in the report, too. But this should be easy, as long as you have the inspectorObject

Best
M0nk3y

**Comment by asterictnl-lvdw** [comment]
@M0nk3yOo ,

Expected and Default values are already in the HTML report where applicable. I will be converting the M365 inspectors first and test them under CUSTOM. A lot of changes have to be implemented regarding this since we are changing the output and validation as well as you want to return all the inspectors now and not only the ones that failed. 

For now I have the following schema:
- UUID (which consists the Unique identifier of the inspector)
- ID (the actual ID which is basically a translation of the UUID with only the numbers e.g. 1.1.1)
- Title (The original title in the CIS Benchmark)
- ProductFamily (Depending on what the audit script says it is placed under a category)
- DefaultValue (Default value if any)
- ExpectedValue (Expected value based on the audit schema)
- ReturnedValue (The actual value which is returned)
- Status (PASS/FAIL/UNKNOWN)
- Notes (Which is blank as you can fill this in yourself)
- RiskScore (is not included in the .CSV)
- RiskRating (is not included in the .CSV)
- Description (which only includes the Rationale and some of the description)
- Impact (which includes the impact or if there is no impact it will be measured)
- Remedation (which is the PowerShell script or the URL if applicable
- References (if any)

Reason that I am not going to implement Partial as Partial is a relative definition which can be adjusted after the audit if the case is that it is partially implemented. But looking at it there is only (implemented or not implemented).

If this is implemented this will be available for testing in the alpha branch first. I will post another comment once it is done.

~LvdW

**Comment by asterictnl-lvdw** [comment]
@M0nk3yOo ,

The first alpha version of the .CSV mechanism is pushed to alpha.

You can execute it with the M365SATTester.ps1 script. This is the command:
`Get-M365SATReport -OutPath "C:\Out" -Username "example@example.org" -EnvironmentType CUSTOM -BenchmarkVersion "Latest" -Modules Office365 -LicenseMode "E3" -LicenseLevel "All" -reportType "CSV" -AllowLogging "Warning" -LocalMode -SkipChecks -ExpirimentalMode`

It is not executing all the inspectors within the paragraph 1 as some are within Exchange as well. Do not change anything in the M365SATTester.ps1 if you want to test the .csv functionality. The full command does not use the experimental mode. The custom inspectors do not work with the normal mode as they have different output formats.

Let me know if you have any feedback regarding the .CSV. I have attached the example_report as an attachment to this comment. 
[example_report.csv](https://github.com/user-attachments/files/17833193/example_report.csv)

~LvdW

**Comment by M0nk3yOo** [comment]
Dear @asterictnl-lvdw,

that looks amazing and will be very very helpfull for us! <3

Here are some feedbacks from my side
- it seems, that there is an issue with the PS-output. CSV says that some checks are successfull, but console says that all failed
<img width="1228" alt="image" src="https://github.com/user-attachments/assets/f23e78a7-582a-4807-8320-7cc0462d0658">
- From my opinion it is very helpfull to see the expected values and the returned one, perfect! I think we can fill it with some more details. For example in 1.1.2 is expected to have at least 2 global admin accounts, and the result (in my test environment) was 3. So it's not equal. Someone could say "hmm... bad for me", but I think the expected value could describe "more than 2" or "> 2" (in correlation with 1.1.3 for sure).
- The CSV file itself (cosmetic): if I open it in Excel, and split it up into columns, the ID (control ID) is broken. Excel interprets it as date. 
<img width="161" alt="image" src="https://github.com/user-attachments/assets/6bed3b1e-c7c1-4d77-939c-b34c751d6755">
Maybe (not the "perfect" solution could be to add a "blank" at the beginning of the ID, so Excel doesn't format it as date (tested in 1.1.1 as you can see)

Question: what If, if one control fails from script perspective (sth. in $error). Will the remediation has the status "ERROR" or "SCRIPTERROR"?


Best & have a nice WE

**Comment by asterictnl-lvdw** [comment]
@M0nk3yOo ,

I gave you the wrong script. Please add an extra parameter called: `-ExpirimentalMode` because it does not work with the normal version.

`Get-M365SATReport -OutPath "C:\Out" -Username "example@example.org" -EnvironmentType CUSTOM -BenchmarkVersion "Latest" -Modules Office365 -LicenseMode "E3" -LicenseLevel "All" -reportType "CSV" -AllowLogging "Warning" -LocalMode -SkipChecks -ExpirimentalMode `

I have fixed it in the M365SATTester.ps1

**Comment by M0nk3yOo** [comment]
Yes, that looks better regarding the results in console.
<img width="870" alt="image" src="https://github.com/user-attachments/assets/8e4a462e-08d0-4c63-8dfe-47b3fafe05ba">


The result of 1.2.1 is missing in CSV

**Comment by asterictnl-lvdw** [comment]
@M0nk3yOo ,

Thank you for reporting.

I have fixed the CISMOFF121 the status should now be correctly displayed. There were some issues returning the correct value. I have tested this, so now the Microsoft365 part level 1 should be working as expected. 

~LvdW

**Comment by asterictnl-lvdw** [comment]
Hi @M0nk3yOo ,

So far I have implemented from the M365 everything but Teams.

Teams will be implemented next week in the .CSV format for testing. Then I will provide the correct command so you can audit everything and check if it is working and displaying all correctly. Then I can also check if I need to make cosmetic changes and such.

If the format and lay-out is correct, then I will proceed converting the AZURE as that one is more challenging.

~LvdW

**Comment by asterictnl-lvdw** [comment]
Hi @olivierdumon and @M0nk3yOo,

I want to inform that all the M365 modules are now converted to support the .CSV format. I need to verify if you guys get the correct output as well. You can run this script:

`Get-M365SATReport -OutPath "C:\Out" -Username "example@example.org" -EnvironmentType CUSTOM -BenchmarkVersion "Latest" -Modules "All" -LicenseMode "All" -LicenseLevel "All" -reportType "CSV" -AllowLogging "Warning" -LocalMode -SkipChecks -ExpirimentalMode`

Make sure to change the username to the desired option. 

Kindly let me know the results.

~LvdW

**Comment by asterictnl-lvdw** [comment]
Okay the last fixes have been applied to the Microsoft 365 Inspectors. I have enhanced the CSV reporting a little bit as well. The .csv works correctly now and outputs all the inspectors with either PASS/FAIL including their values etc. You can test it with the script above. 

Do not change the EnvironmentType or remove the ExpirimentalMode parameter as that will break the script!!!

The next part would be that I need to implement any changes in the HTML report schema as well. 

Then I have the convert the Azure/Entra inspectors as well to support the .CSV format.

But for you can audit based on the Microsoft 365 ecosystem or use the classic .HTML which supports both. 

Happy Auditing :)

~LvdW

**Comment by M0nk3yOo** [comment]
Hi @asterictnl-lvdw 

thanks for all the updates and work!
I'm doing some tests in parallel of my work, but faced some issues which I wanted to doublecheck.

To got a note that there was also a new CIS-version published this week? So far I read till now some inspecores have been removed, and some added.


Have a nice weekend!
M0nk3y

**Comment by asterictnl-lvdw** [comment]
@M0nk3yOo ,

v4.0.0 is included in the alpha branch. You can use it there. I want to release the .CSV compatibility along with all inspectors fully converted. For the rest I saw there is no update released yet. I am looking at the workbench carefully to determine if updates are released and when I have to check and implement the new versions of the inspectors.

So far, the latest versions are v4.0.0 for M365 and v3.1.0 for Azure. 

~LvdW

**Comment by M0nk3yOo** [comment]
Hi @asterictnl-lvdw 

do you have the same problem as me? I already updated the modules (tested with PS7)

<img width="1012" alt="image" src="https://github.com/user-attachments/assets/22b283cc-09ab-4e7f-a377-85979268afd7" />

Best

**Comment by asterictnl-lvdw** [comment]
Hello @M0nk3yOo ,

Please remove all modules and install them one-by-one again to see if the issue pre-exists. 

I will test for you and come back with results.

~LvdW

**Comment by asterictnl-lvdw** [comment]
@M0nk3yOo ,

This is an issue regarding WAM and the new login experience that has been released in v3.7.0. You can find details about that Changelog here. I am going to file a ticket to Microsoft so they can add a parameter that disables the WAMLoginExperience.

https://blog.icewolf.ch/archive/2024/12/03/exchangeonlinemanagement-3-7-0-released/

A workaround is to downgrade the module to 3.6.0 until further notice as that module does not use the WAM mechanism and works with the flow. 

I will keep you updated about this.

~LvdW

**Comment by asterictnl-lvdw** [comment]
Happy New Year everyone,

I just want to let you guys know that I scheduled the migration of the Azure inspectors to support the .CSV format. The works for that will hopefully be done before the end of Q1. After that there will be changes to the HTML report too to support the new inspector format.

~LvdW

**Comment by M0nk3yOo** [comment]
Hi, and I wish you a happy new year, too! Thanks :-)

Also Thanks for your update. I still need to test/check my environment... :-/

**Comment by olivierdumon** [comment]
> Happy New Year everyone,
> 
> I just want to let you guys know that I scheduled the migration of the Azure inspectors to support the .CSV format. The works for that will hopefully be done before the end of Q1. After that there will be changes to the HTML report too to support the new inspector format.
> 
> ~LvdW

Happy new year, too
Thanks again for all your work.
I'll take time to test the latest updates
Regards

**Comment by M0nk3yOo** [comment]
Hi all,


I've reinstalled the ExchangeOnlineModule with success, but I think the problem (or my personal problem?) is more related to the MSAL authentication:

<img width="1023" alt="image" src="https://github.com/user-attachments/assets/151ab0d2-228f-402c-b7b6-7583651a74ab" />

Maybe it's related to the virtual ARM-Version (I've currently a Mac and have Windows running in an VM via "VMware Fusion" but it was working before. It could also be another module which was better to not update. I always update all modules in one run from time to time...
If anything is working on your side, I'll try another VM x86 based VM on Proxmox

**Comment by asterictnl-lvdw** [comment]
@M0nk3yOo ,

Please remove all modules in the respective PowerShell/Modules folder and reinstall them and make sure to install this ExchangeOnlineManagement version:

`Install-Module -Name ExchangeOnlineManagement -RequiredVersion 3.6.0`

Also make sure when you authenticate keep the localhost sessions open as they are essential to keep the session up. They can be closed afterwards. This only occurs at PowerShell 7, because PowerShell 5 uses a different mechanism to authenticate.

Then it should work.

~LvdW

**Comment by M0nk3yOo** [comment]
Dear @asterictnl-lvdw ,

I've now a fresh installed win 11 PC and installed the recommendet modules and PS7 as well.
PS5.1 and 7.5 behaviours a bit different and don't know why.

PS5.1:
Installed modules:

![Image](https://github.com/user-attachments/assets/1d4c0f2f-49a7-4855-ab1a-42bcb3839738)

Error:

![Image](https://github.com/user-attachments/assets/4f9c245f-ae2a-4399-82d4-6d79fa3bcee7)



PS7.5
Installed modules:

![Image](https://github.com/user-attachments/assets/81d8702f-366e-4ef9-a017-775d85ccc1f6)

Error1 (first as in PS5.1)

![Image](https://github.com/user-attachments/assets/b0b2b5a0-9bb9-4189-a144-896f68b0c076)

Error2

![Image](https://github.com/user-attachments/assets/f8195726-f00a-4f7e-bffa-e77ff8eee30a)
I can confirm that Tab-completion for "Connect-MicrosoftTeams" is working in 5.1 and not in 7.1



So currently it seams that PS5.1 brings the best results, with the old modules.



I also got some inspector errors. Should I paste them here, or should I create different issues?

**Comment by asterictnl-lvdw** [comment]
Hi @M0nk3yOo ,

On my Linux machine the LoginExperienceV2 works without any problems on both Linux and Windows 11 x64.

Use these commands to install properly in this order in PowerShell 7:
Install-Module -Name Az
Install-Module -Name ExchangeOnlineManagement -RequiredVersion 3.6.0
Install-Module -Name Microsoft.Graph -AllowClobber -Force
Install-Module -Name Microsoft.Graph.Beta -AllowClobber -Force
Install-Module -Name MicrosoftTeams
Install-Module -Name PoShLog

Install this command in PowerShell 5:
Install-Module -Name Microsoft.Online.SharePoint.PowerShell

Please check if this command is working separately, you should get an output that it is disabled:

`Set-AzConfig -EnableLoginByWam $false -LoginExperienceV2 'Off'`

If that command works then you should be fine.

Please install everything in PowerShell 7 except the Microsoft Sharepoint module which needs to be installed in PowerShell 5 in order to make it work. If you install things in PowerShell 5 it sometimes does not work for PowerShell 7 as the profile is different and you need to change some settings to add the path to the profile, but that could cause conflicts since those modules might not be designed for PowerShell 7.


For the inspector errors please create a seperate issue for this so I can look into it.

Thank you.

~LvdW

**Comment by asterictnl-lvdw** [comment]
@M0nk3yOo ,

Thank you for sharing the screenshot. I will update the command so the alias is not used. That would avoid issues with others who did not manage to register the alias properly.

Can you please run the M365SATConnectionTester in the alpha branch and show me if there are any issues?

If so then I have to open a seperate issue for the MicrosoftTeams module. 

~LvdW

**Comment by Karmakstylez** [comment]
Dear @M0nk3yOo  and @olivierdumon ,
All inspectors from both Azure and M365 are converted right now. I will need to convert some scripts and the .HTML format as well and then I can release v3.1 where .CSV is supported for both formats. Once that is done you can test it in the alpha branch. I plan to release the stable version of v3.1 in March if everything goes well.

~LvdW

**Comment by Karmakstylez** [comment]
The .CSV version has been successfully released. I will close this issue. 

If there are any issues with the .CSV or inspectors feel free to open a different issue.

~LvdW
